src/levanter/models/whisper.py

```diff
@@ -0,0 +1,497 @@
+
+    @named_call
+    def __call__(
+        self, x: NamedArray, xa: Optional[NamedArray] = None, mask: Optional[AttentionMask | NamedArray] = None, *, key
+    ):
+        k1, k2, k3 = haliax.jax_utils.maybe_rng_split(key, 3)
+
+        attn_output = self.attn(self.attn_ln(x), mask=mask, key=k1)
+        x = x + attn_output
+
+        if self.cross_attn and self.cross_attn_ln:
+            cross_attn_output = self.cross_attn(self.cross_attn_ln(x), xa, key=k2)
+            x = x + cross_attn_output
+
+        ff_output = self.mlp(self.mlp_ln(x), key=k3)
+        x = x + ff_output
+
+        return x
+
+
+class WhisperTransformer(eqx.Module):
+    blocks: Stacked[WhisperBlock]
+    Layers: Axis
+    ln_f: hnn.LayerNorm
+
+    @staticmethod
+    def init(Layers: Axis, Heads: Axis, HeadSize: Axis, config: WhisperConfig, has_cross: bool, *, key):
+        # vectorize the blocks
+        blocks = Stacked.init(Layers, WhisperBlock, gradient_checkpointing=config.gradient_checkpointing)(
+            Heads,
+            HeadSize,
+            config,
+            has_cross=has_cross,
+            key=shaped_rng_split(key, Layers.size),
+        )
+        ln_f = hnn.LayerNorm.init(config.Embed, eps=config.layer_norm_epsilon, use_bias=config.use_bias)
+
+        return WhisperTransformer(blocks, Layers, ln_f)
+
+    @named_call
+    def __call__(
+        self,
+        x: NamedArray,
+        xa: Optional[NamedArray] = None,
+        attn_mask: Optional[AttentionMask | NamedArray] = None,
+        *,
+        key=None,
+    ) -> NamedArray:
+        keys = hax.jax_utils.maybe_rng_split(key, self.Layers) if key is not None else None
+        x = self.blocks.fold(x, xa, attn_mask, key=keys)
+        x = self.ln_f(x)
+
+        return x
+
+
+class WhisperSinusoidEmbedding(eqx.Module, StateDictSerializationMixin):
+    Pos: Axis = eqx.field(static=True)
+    pos_emb: NamedArray
+
+    @classmethod
+    def init(cls, Channels: Axis, SourcePos: Axis, base: int = 10000):
+        # this must be compile-time b/c we want to store them in a static field
+        with jax.ensure_compile_time_eval():
+            pos_emb = cls._get_pos_emb(SourcePos=SourcePos, Channels=Channels, base=base)
+
+        return WhisperSinusoidEmbedding(SourcePos, pos_emb)
+
+    @staticmethod
+    def _get_pos_emb(Channels: Axis, SourcePos: Axis, base: int) -> NamedArray:
+        log_timescale_increment = (hax.log(base) / (Channels.size // 2 - 1)).item()
+        ChannelHalfSize = Channels.resize(Channels.size // 2)
+        inv_timescales = hax.exp(-log_timescale_increment * hax.arange(ChannelHalfSize))
+
+        position_ids: NamedArray = hax.arange(SourcePos)
+
+        freqs = position_ids * inv_timescales.broadcast_axis(SourcePos)
+        emb = hax.concatenate(Channels, (hax.sin(freqs), hax.cos(freqs)))
+        return emb
+
+    def __call__(self, seq_len: int) -> NamedArray:
+        return jax.lax.stop_gradient(self.pos_emb[self.Pos, :seq_len])
```

Comment:
this is unfortunately not enough (by itself) to stop the embedding from being modified (namely by weight decay). This being JAX, you're best off just making this a def and using compite_time_eval in the jit block. (I just realized this today) thttps://github.com/stanford-crfm/levanter/pull/503/files

Reply:
Perfect, adapted my version to use the compile_time_eval block modeled after the LLama one!